dm/syncer/validate_worker.go

```diff
@@ -0,0 +1,348 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"github.com/pingcap/tidb/parser/model"
+	tidbmysql "github.com/pingcap/tidb/parser/mysql"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+	"github.com/pingcap/tiflow/dm/syncer/dbconn"
+)
+
+const (
+	workerChannelSize = 1000
+)
+
+type validateWorker struct {
+	cfg               config.ValidatorConfig
+	ctx               context.Context
+	interval          time.Duration
+	validator         *DataValidator
+	L                 log.Logger
+	conn              *dbconn.DBConn
+	rowChangeCh       chan *rowChange
+	pendingChangesMap map[string]*tableChange
+	pendingRowCount   atomic.Int64
+	sync.Mutex
+
+	// used for test
+	receivedRowCount atomic.Int64 // number of rowChange received from channel
+	validationCount  atomic.Int64 // number of successful validation
+}
+
+func newValidateWorker(v *DataValidator, id int) *validateWorker {
+	return &validateWorker{
+		cfg:               v.cfg.ValidatorCfg,
+		ctx:               v.ctx,
+		interval:          v.validateInterval,
+		validator:         v,
+		L:                 v.L,
+		conn:              v.toDBConns[id],
+		rowChangeCh:       make(chan *rowChange, workerChannelSize),
+		pendingChangesMap: make(map[string]*tableChange),
+	}
+}
+
+func (vw *validateWorker) run() {
+outer:
+	for {
+		select {
+		case change, ok := <-vw.rowChangeCh:
+			if !ok {
+				break outer
+			}
+			// todo: limit number of pending rows
+			vw.updateRowChange(change)
+			vw.receivedRowCount.Inc()
+		case <-time.After(vw.interval):
+			err := vw.validateTableChange()
```

Comment:
Can the validation be triggered by the number of pending row? What if the worker gets too much `rowChange` in a short period? Setting a batch threshold can restrict the resource upper bound that the worker consumes at a time.

Reply:
add a todo to limit number of pending rows here

Reply:
add a todo to limit number of pending rows here

https://github.com/buchuitoudegou/tiflow/blob/ebba8bb19d0083a8653b815617ff55cea4a50fe8/dm/syncer/validate_worker.go#L82

might be better to trigger validation at that time, will leave a todo.

but what if the validation didn't reduce the number of pending row change(suppose most rows are changed again), and retry might not work then, so we have to keep accumulate row changes until we approach progress of syncer(row changes get updated and most row change can validated successfully)